File: examples/latest/navigate/flutter/map_items_app/lib/MapObjectsExample.dart
/*
 * Copyright (C) 2019-2022 HERE Europe B.V.
 *
 * Licensed under the Apache License, Version 2.0 (the "License")
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * SPDX-License-Identifier: Apache-2.0
 * License-Filename: LICENSE
 */

import 'package:flutter/material.dart';
import 'package:here_sdk/core.dart';
import 'package:here_sdk/core.errors.dart';
import 'package:here_sdk/mapview.dart';

class MapObjectsExample {
  final double _distanceInMeters = 1000 * 10;
  final GeoCoordinates _berlinGeoCoordinates = new GeoCoordinates(52.51760485151816, 13.380312380535472);

  final MapScene _mapScene;
  final MapCamera _mapCamera;
  MapPolyline? _mapPolyline;
  MapArrow? _mapArrow;
  MapPolygon? _mapPolygon;
  MapPolygon? _mapCircle;

  MapObjectsExample(HereMapController hereMapController)
      : _mapScene = hereMapController.mapScene,
        _mapCamera = hereMapController.camera {
    double distanceToEarthInMeters = 5000;
    hereMapController.camera
        .lookAtPointWithDistance(GeoCoordinates(52.51760485151816, 13.380312380535472), distanceToEarthInMeters);
  }

  void showMapPolyline() {
    clearMap();
    // Move map to expected location.
    _mapCamera.flyToWithOptionsAndDistance(
        _berlinGeoCoordinates, _distanceInMeters, new MapCameraFlyToOptions.withDefaults());

    _mapPolyline = _createPolyline();
    _mapScene.addMapPolyline(_mapPolyline!);
  }

  void showMapArrow() {
    clearMap();
    // Move map to expected location.
    _mapCamera.flyToWithOptionsAndDistance(
        _berlinGeoCoordinates, _distanceInMeters, new MapCameraFlyToOptions.withDefaults());

    _mapArrow = _createMapArrow();
    _mapScene.addMapArrow(_mapArrow!);
  }

  void showMapPolygon() {
    clearMap();
    // Move map to expected location.
    _mapCamera.flyToWithOptionsAndDistance(
        _berlinGeoCoordinates, _distanceInMeters, new MapCameraFlyToOptions.withDefaults());

    _mapPolygon = _createPolygon();
    _mapScene.addMapPolygon(_mapPolygon!);
  }

  void showMapCircle() {
    clearMap();
    // Move map to expected location.
    _mapCamera.flyToWithOptionsAndDistance(
        _berlinGeoCoordinates, _distanceInMeters, new MapCameraFlyToOptions.withDefaults());

    _mapCircle = _createMapCircle();
    _mapScene.addMapPolygon(_mapCircle!);
  }

  void clearMap() {
    if (_mapPolyline != null) {
      _mapScene.removeMapPolyline(_mapPolyline!);
    }

    if (_mapArrow != null) {
      _mapScene.removeMapArrow(_mapArrow!);
    }

    if (_mapPolygon != null) {
      _mapScene.removeMapPolygon(_mapPolygon!);
    }

    if (_mapCircle != null) {
      _mapScene.removeMapPolygon(_mapCircle!);
    }
  }

  MapPolyline? _createPolyline() {
    List<GeoCoordinates> coordinates = [];
    coordinates.add(GeoCoordinates(52.53032, 13.37409));
    coordinates.add(GeoCoordinates(52.5309, 13.3946));
    coordinates.add(GeoCoordinates(52.53894, 13.39194));
    coordinates.add(GeoCoordinates(52.54014, 13.37958));

    GeoPolyline geoPolyline;
    try {
      geoPolyline = GeoPolyline(coordinates);
    } on InstantiationException {
      // Thrown when less than two vertices.
      return null;
    }

    double widthInPixels = 20;
    Color lineColor = Color.fromARGB(160, 0, 144, 138);
    MapPolyline mapPolyline = MapPolyline(geoPolyline, widthInPixels, lineColor);

    return mapPolyline;
  }

  MapArrow? _createMapArrow() {
    List<GeoCoordinates> coordinates = [];
    coordinates.add(GeoCoordinates(52.53032, 13.37409));
    coordinates.add(GeoCoordinates(52.5309, 13.3946));
    coordinates.add(GeoCoordinates(52.53894, 13.39194));
    coordinates.add(GeoCoordinates(52.54014, 13.37958));

    GeoPolyline geoPolyline;
    try {
      geoPolyline = GeoPolyline(coordinates);
    } on InstantiationException {
      // Thrown when less than two vertices.
      return null;
    }

    double widthInPixels = 20;
    Color lineColor = Color.fromARGB(160, 0, 144, 138);
    MapArrow mapArrow = MapArrow(geoPolyline, widthInPixels, lineColor);

    return mapArrow;
  }

  MapPolygon? _createPolygon() {
    List<GeoCoordinates> coordinates = [];
    // Note that a polygon requires a clockwise or counter-clockwise order of the coordinates.
    coordinates.add(GeoCoordinates(52.54014, 13.37958));
    coordinates.add(GeoCoordinates(52.53894, 13.39194));
    coordinates.add(GeoCoordinates(52.5309, 13.3946));
    coordinates.add(GeoCoordinates(52.53032, 13.37409));

    GeoPolygon geoPolygon;
    try {
      geoPolygon = GeoPolygon(coordinates);
    } on InstantiationException {
      // Less than three vertices.
      return null;
    }

    Color fillColor = Color.fromARGB(160, 0, 144, 138);
    MapPolygon mapPolygon = MapPolygon(geoPolygon, fillColor);

    return mapPolygon;
  }

  MapPolygon _createMapCircle() {
    double radiusInMeters = 300;
    GeoCircle geoCircle = GeoCircle(GeoCoordinates(52.51760485151816, 13.380312380535472), radiusInMeters);

    GeoPolygon geoPolygon = GeoPolygon.withGeoCircle(geoCircle);
    Color fillColor = Color.fromARGB(160, 0, 144, 138);
    MapPolygon mapPolygon = MapPolygon(geoPolygon, fillColor);

    return mapPolygon;
  }
}

File: examples/latest/lite/ios/MapItemsLite/MapItemsLite/MapObjectsExample.swift
/*
 * Copyright (C) 2019-2022 HERE Europe B.V.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * SPDX-License-Identifier: Apache-2.0
 * License-Filename: LICENSE
 */

import heresdk

class MapObjectsExample {

    private let berlinGeoCoordinates = GeoCoordinates(latitude: 52.51760485151816, longitude: 13.380312380535472)
    
    private let mapScene: MapSceneLite
    private let mapCamera: CameraLite
    private var mapPolyline: MapPolylineLite?
    private var mapPolygon: MapPolygonLite?
    private var mapCircle: MapCircleLite?

    init(mapView: MapViewLite) {
        // Configure the map.
        mapCamera = mapView.camera
        mapCamera.setTarget(GeoCoordinates(latitude: 52.51760485151816, longitude: 13.380312380535472))
        mapCamera.setZoomLevel(13)

        mapScene = mapView.mapScene
    }

    func onMapPolylineClicked() {
        clearMap()
        // Move map to expected location.
        mapCamera.setTarget(berlinGeoCoordinates)
        mapCamera.setZoomLevel(13.0)
        
        mapPolyline = createMapPolyline()
        mapScene.addMapPolyline(mapPolyline!)
    }

    func onMapPolygonClicked() {
        clearMap()
        // Move map to expected location.
        mapCamera.setTarget(berlinGeoCoordinates)
        mapCamera.setZoomLevel(13.0)
        
        mapPolygon = createMapPolygon()
        mapScene.addMapPolygon(mapPolygon!)
    }

    func onMapCircleClicked() {
        clearMap()
        // Move map to expected location.
        mapCamera.setTarget(berlinGeoCoordinates)
        mapCamera.setZoomLevel(13.0)
        
        mapCircle = createMapCircle()
        mapScene.addMapCircle(mapCircle!)
    }

    func onClearButtonClicked() {
        clearMap()
    }

    private func createMapPolyline() -> MapPolylineLite {
        let coordinates = [GeoCoordinates(latitude: 52.53032, longitude: 13.37409),
                           GeoCoordinates(latitude: 52.5309, longitude: 13.3946),
                           GeoCoordinates(latitude: 52.53894, longitude: 13.39194),
                           GeoCoordinates(latitude: 52.54014, longitude: 13.37958)]

        // We are sure that the number of vertices is greater than two, so it will not crash.
        let geoPolyline = try! GeoPolyline(vertices: coordinates)
        let mapPolylineStyle = MapPolylineStyleLite()
        mapPolylineStyle.setWidthInPixels(inPixels: 20.0)
        mapPolylineStyle.setColor(0x00908AA0, encoding: .rgba8888)
        let mapPolyline = MapPolylineLite(geometry: geoPolyline, style: mapPolylineStyle)

        return mapPolyline
    }

    private func createMapPolygon() -> MapPolygonLite {
        let coordinates = [GeoCoordinates(latitude: 52.53032, longitude: 13.37409),
                           GeoCoordinates(latitude: 52.5309, longitude: 13.3946),
                           GeoCoordinates(latitude: 52.53894, longitude: 13.39194),
                           GeoCoordinates(latitude: 52.54014, longitude: 13.37958)]

        // We are sure that the number of vertices is greater than three, so it will not crash.
        let geoPolygon = try! GeoPolygon(vertices: coordinates)
        let mapPolygonStyle = MapPolygonStyleLite()
        mapPolygonStyle.setFillColor(0x00908AA0, encoding: .rgba8888)
        let mapPolygon = MapPolygonLite(geometry: geoPolygon, style: mapPolygonStyle)

        return mapPolygon
    }

    private func createMapCircle() -> MapCircleLite {
        let geoCircle = GeoCircle(center: GeoCoordinates(latitude: 52.51760485151816, longitude: 13.380312380535472),
                                  radiusInMeters: 300.0)
        let mapCircleStyle = MapCircleStyleLite()
        mapCircleStyle.setFillColor(0x00908AA0, encoding: .rgba8888)
        let mapCircle = MapCircleLite(geometry: geoCircle, style: mapCircleStyle)

        return mapCircle
    }

    private func clearMap() {
        if let line = mapPolyline {
            mapScene.removeMapPolyline(line)
        }

        if let area = mapPolygon {
            mapScene.removeMapPolygon(area)
        }

        if let circle = mapCircle {
            mapScene.removeMapCircle(circle)
        }
    }
}
